### PR TITLE
Allow persistent resizing of the widget app drawer

### PR DESCRIPTION
**Pull request by t3chguy**
Supercedes https://github.com/matrix-org/matrix-react-sdk/pull/4493
Fixes https://github.com/vector-im/element-web/issues/6785

Demo at https://riots.im/adhoc/dpsah-widgets-resizing4/

### CODE REVIEW

File: res/css/views/rooms/_AppsDrawer.scss [hunk 1]
```diff
@@ -358,3 +411,32 @@ form.mx_Custom_Widget_Form div {
...
+    position: relative;
+    z-index: 1;
+}
+
+.mx_AppsDrawer .mx_ResizeHandle > div {
+    background: inherit;
+}
+
+.mx_AppsDrawer_fullWidth .mx_ResizeHandle {
+    max-width: 960px;
```
Comment (jryans) [inline res/css/views/rooms/_AppsDrawer.scss:426]: Why this particular number?

File: src/hooks/useLocalStorage.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {useEffect, useRef, useState} from "react";
+
+// Hook behaving like useState but persisting the value to localStorage. Returns same as useState
+export const useLocalStorageState = (key: string, initialValue: boolean) => {
+    const lsKey = useRef("useLocalStorageState_" + key).current;
```
Comment (jryans) [inline src/hooks/useLocalStorage.ts:21]: Why do you need the `useRef` hook here? This is to track changes to the key...?

About the item name that's used in local storage, adding a prefix of `useLocalStorageState_` seems to bake in an implementation detail into the data being stored. Maybe use a generic `mx_` prefix, or just leave it up to the caller?

**Review by jryans — CHANGES_REQUESTED** [review]
Overall looks reasonable, just a few questions.

File: src/hooks/useLocalStorage.ts (hunk 2 above)
Comment (t3chguy) [reply src/hooks/useLocalStorage.ts:21]: The hook here prevents the value from ever changing, it takes the initial value and stores it in the ref

> seems to bake in an implementation detail into the data being stored.

given that it serializes + deserializes it is implementation specific, no?

Comment (t3chguy) [reply src/hooks/useLocalStorage.ts:21]: Instead I could make it save the old key and load the value form the new key on a change, that would be more re-usable

Comment (jryans) [reply src/hooks/useLocalStorage.ts:21]: > given that it serializes + deserializes it is implementation specific, no?

Hmm, I don't know if that part is really that specific to exactly this function, it's fairly common to do that with local storage. 

The main thing I am reacting to is the name `useLocalStorageState` has little / nothing to do with the data being stored, so it's a bit surprising to embed that in key name being stored. Maybe later we'll rewrite this code into a function with a different name, but we'd still have a bunch of keys stored in local storage with the prefix `useLocalStorageState_`. The name of a specific data access function that happens to exist at the moment seems like too specific and too temporary a thing to place in a storage layer to me.

File: src/hooks/useLocalStorageState.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2019 The Matrix.org Foundation C.I.C.
```
Comment (jryans) [inline src/hooks/useLocalStorageState.ts:2]:
```suggestion
Copyright 2020 The Matrix.org Foundation C.I.C.
```

**Review by jryans — APPROVED** [review]
Thanks, looks good to me! 😄

**Review by nadonomy — CHANGES_REQUESTED** [review]
@t3chguy is there anything we can do to make the timeline position persist while resizing? It feels really strange now where resizing widgets to increase their height also impacts the timeline view, where it'd be better if the scroll position persisted.

**Comment by t3chguy** [comment]
So the magical timeline scroll behaviour should already be doing that. I'm afraid tweaking it is more likely to break normal behaviour, it is a very complex thing.
I can make it work for when you are resizing slowly but when you speed up it breaks

**Comment by t3chguy** [comment]
One way would be to have the AppsDrawer float atop the timeline so it never causes it to resize which will likely make resizing more performant too

File: src/components/views/elements/PersistedElement.js [hunk 4]
```diff
@@ -167,9 +167,9 @@ export default class PersistedElement extends React.Component {
             width: parentRect.width + 'px',
             height: parentRect.height + 'px',
         });
-    }
+    }, 100, {trailing: true, leading: true});
```
Comment (bwindels) [inline src/components/views/elements/PersistedElement.js:170]: Curious, why did this need to be added?

**Review by bwindels — APPROVED** [review]
Have looked at scroll panel and resize notifier changes, and that looks :+1:

File: src/components/views/elements/PersistedElement.js (hunk 4 above)
Comment (t3chguy) [reply src/components/views/elements/PersistedElement.js:170]: This is just so the PersistedElement (floating iframe) doesn't update too frequently, it was causing the whole resize operation to lag

**Review by nadonomy — APPROVED** [review]
Feeling good, thanks!

**Pull request merged**